vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,797 @@
+                | None ->
+                    // This function hasn't got any cache and so it's completely new.
+                    // So create completely new results
+                    // And finally add a tag for this.
+                    let res = 
+                        CodeLens(
+                            Async.cache (useResults (symbolUse.DisplayContext, func)),
+                            false,
+                            fullTypeSignature,
+                            null)
+                    try
+                        let declarationSpan = textSnapshot.GetLineFromLineNumber(func.DeclarationLocation.StartLine - 1).Extent.Span
+                        let trackingSpan = 
+                            textSnapshot.CreateTrackingSpan(declarationSpan, SpanTrackingMode.EdgeInclusive)
+                        codeLensToAdd.Add (trackingSpan, res)
+                        newResults.[fullDeclarationText] <- (trackingSpan, res)
+                    with e -> logExceptionWithContext (e, "Code Lens tracking tag span creation")
+                ()
+
+            //for tagToUpdate in tagsToUpdate do
+            //    do! removeTag(tagToUpdate.Key)
+            //    let fullDeclarationText, tag = tagToUpdate.Value
+            //    let! newTag = addTag (tagToUpdate.Key.Span) (CodeLensTag(0., GetTextBlockSize.Height, 0., 0., 0., PositionAffinity.Predecessor, tag, self))
+            //    newResults.[fullDeclarationText] <- newTag
+            lastResults <- newResults
+            do! Async.SwitchToContext uiContext |> liftAsync
+            
+            self.Mutex.WaitOne() |> ignore
+            for value in codeLensToAdd do
+                let trackingSpan, codeLens = value
+                let stackPanel = self.AddCodeLens trackingSpan
+                logInfof "Trackingspan %A is being added." trackingSpan 
+                stackPanel.IsVisibleChanged
+                |> Event.filter (fun eventArgs -> eventArgs.NewValue :?> bool)
+                |> Event.add (fun _ ->
+                    if codeLens.Computed |> not then
+                        async {
+                            do! createTextBox codeLens
+                            let uiElement = codeLens.UiElement
+                            do! Async.SwitchToContext uiContext
+                            self.AddUiElementToCodeLensOnce trackingSpan uiElement
+                            logInfo "Adding text box!"
+                        } |> RoslynHelpers.StartAsyncSafe CancellationToken.None
+                    )
+
+            for oldResult in oldResults do
+                let trackingSpan, _ = oldResult.Value
+                logInfof "removing trackingSpan %A" trackingSpan
+                self.RemoveCodeLens trackingSpan |> ignore
+
+            logInfof "Finished updating code lens."
+            
+            if not firstTimeChecked then
+                firstTimeChecked <- true
+            self.Mutex.ReleaseMutex()
```

Comment:
What if there is an exception in the Mutex region?  SHould the ``Release`` be part of a ``try/finally``, i.e. governed by a ``use disposer = ...``